packages/mcfunction/src/node/macro.ts

```diff
@@ -0,0 +1,66 @@
+import * as core from '@spyglassmc/core'
+
+export interface MacroNode extends core.SequenceNode<MacroChildNode> {
+	type: 'mcfunction:macro'
+	slash?: core.Range
+	children: MacroChildNode[]
+}
+export namespace MacroNode {
+	/* istanbul ignore next */
+	export function is<T extends core.DeepReadonly<core.AstNode> | undefined>(
+		obj: T,
+	): obj is core.InheritReadonly<MacroNode, T> {
+		return (obj as MacroNode | undefined)?.type ===
+			'mcfunction:macro'
+	}
+
+	export function mock(range: core.RangeLike): MacroNode {
+		return {
+			type: 'mcfunction:macro',
+			range: core.Range.get(range),
+			children: [MacroChildNode.mock(range)],
+		}
+	}
+}
+
+export interface MacroOptions {
+	type?: 'macro' | 'other' | 'sign'
+	colorTokenType?: core.ColorTokenType
```

Comment:
Storing the color token in the AST feels wrong. The AST is supposed to represent the structure of the document. I haven't reviewed the rest of the PR yet, but imo a command like `$say $(message)` should look a bit like this (simplified AST):
```yaml
type: mcfunction:macro
children:
- type: mcfunction:macro_start
  value: "$"
- type: mcfunction:macro_raw
  value: 'say '
- type: mcfunction:macro_key
  children:
  - type: string
    value: message
```

Reply:
It seems like there's precedent for this in lots of other nodes?

Reply:
The colorTokenType code was taken from `LiteralNode` and it's done the same way there. As for storing the type it's done more for organizational reasons and it's done the same way for `ResourceLocation`s

Reply:
The colorizer would look at the type of the nodes and assign colors based on that. Color tokens in nodes are currently only used for some literal types. Not for distinquishing between the color of parts like these in a command: `$(`, `say `, `message`. Those shouldn't all have the same `macro_child` node type.

Reply:
Actually we came up with a way to improve upon this, once its pushed hopefully that's an adjustment that can resolve this

Reply:
I've removed the `colorTokenType` parameter and just the nodes are now just colored based on their type (which is just literal or string)

Reply:
Looks like an improvement, but the `options.type` just seems like a secondary type. Why do they share the same `macro_child` node type instead of being 3 different AST node types?